homeassistant/components/websocket_api/commands.py

```diff
@@ -231,14 +232,53 @@ async def handle_call_service(
         connection.send_result(msg["id"], {"context": context})
     except ServiceNotFound as err:
         if err.domain == msg["domain"] and err.service == msg["service"]:
-            connection.send_error(msg["id"], const.ERR_NOT_FOUND, "Service not found.")
+            connection.send_error(
+                msg["id"],
+                const.ERR_NOT_FOUND,
+                f"Service {err.domain}.{err.service} not found.",
+                translation_domain=err.translation_domain,
+                translation_key=err.translation_key,
+                translation_placeholders=err.translation_placeholders,
+            )
         else:
-            connection.send_error(msg["id"], const.ERR_HOME_ASSISTANT_ERROR, str(err))
+            # The called service called another service which does not exist
+            connection.send_error(
+                msg["id"],
+                const.ERR_HOME_ASSISTANT_ERROR,
+                f"Service {err.domain}.{err.service} called service "
+                f"{msg['domain']}.{msg['service']} which was not found.",
+                translation_domain=const.DOMAIN,
+                translation_key="child_service_not_found",
+                translation_placeholders={
+                    "domain": err.domain,
+                    "service": err.service,
+                    "child_domain": msg["domain"],
+                    "child_service": msg["service"],
+                },
+            )
     except vol.Invalid as err:
         connection.send_error(msg["id"], const.ERR_INVALID_FORMAT, str(err))
+    except ServiceValidationError as err:
+        connection.logger.error(err)
+        connection.logger.debug("", exc_info=err)
+        connection.send_error(
+            msg["id"],
+            const.ERR_HOME_ASSISTANT_ERROR,
```

Comment:
Why do we use the same `code` for all different exceptions, besides `vol.Invalid`? How does the frontend use the `code` parameter?

Reply:
I used `ERR_HOME_ASSISTANT_ERROR` because `ServiceNotFound` is a subclass of HomeAssistantError and the frontend does not seem to use the error code. But I could make one for `ServiceValidationError` I guess.

Reply:
I'm not sure. I think we need to define what this parameter means and frame it to know what the value should be.

Reply:
I'll make a seperate code for it. I checked, but it is not used in the frontend yet. Only in a special case it seems to be used:
```javascript
try {
      await fetchCloudAlexaEntity(this.hass, this.entityId);
    } catch (err: any) {
      if (err.code === "not_supported") {
        this._unsupported["cloud.alexa"] = true;
        this.requestUpdate("_unsupported");
      }
    }
```

Reply:
> I'm not sure. I think we need to define what this parameter means and frame it to know what the value should be.

If has to have a direct relation with the frontend code, then it seems only `not_supported` has that.
When we are actually redesigning the frontend UX, then the code could be useful the classify.
If we want to define what this parameter means, do you think that should effect this PR? Or is this something to document / rework in a follow up PR?

Reply:
I think it's very closely related to the rest of the changes, so it would be good to think about this now.

Reply:
@MartinHjelmare can you clarify a bit what you mean? Do you mean adding translations to exceptions is related to error codes sent when a websocket API request fails?

Websocket API error responses already have:
- `code`: an error code
- `message`: a human readable error message
With this PR we also add:
- A translatable human readable exception message

I think the biggest overlap is between `message` and the translatable translation, maybe `message` should be `None` when the error has a translation?

Reply:
We're changing how we send errors. It's good to know what the existing parameters of that api means.

Reply:
The current definition of the error `code` is a string that should be unique per error per websocket command, ie not globally unique. We should be able to distinguish errors should using the error code for a websocket command.

Reply:
I suppose we can answer the discussion https://github.com/home-assistant/architecture/discussions/992 now

Reply:
I don't think it's a reasonable requirement that for each possible thing that can go wrong in core when handling a websocket command there's a unique `code`,  but maybe that's not what you mean @MartinHjelmare?

Also, we should clarify the relation between `mesage` and translatable error message. Are both needed, or should `mesage` be `None` when there's a translatable error message?

Reply:
The codes don't have to be globally unique as I wrote above. Per command it should be possible to distinguish different errors that are returned and the same error should always use the same code for that command.

Reply:
It's up to the developer to decide what error to return when something goes wrong, ie what error the thing going wrong should be labeled as.